### DISCUSSION TITLE
Make '-vv..' useful (again)

### DISCUSSION THREAD

**Discussion by icing** [curl/curl#13810] [Ideas]
There is an idea of making repeated `-v` curl command line increase the verbosity of the trace output (again, see below). Personally, I'd find that quite useful as I often need more information when analysing problems. I start this discussion for people to give feedback on my proposal below. Hopefully, we can reach a "rough consensus" and our BDFL can put his finger in the wind to reach a conclusion.

### The Idea

Repeated `-v` should not do anything that cannot be achieved by other, explicit options. It should be a shortcut for them to improve usability. curl has `--trace-ids`, `--trace-time` and the general `--trace-config features`. Then there are `--trace file` and `--trace-ascii file`. I often find myself adding `--trace-config ids,time,http/2,http/3` in my debugging attempts (or asking issue reporters to do this) and that is a bit cumbersome.

The proposed change is to match `-vv` and `-vvv` to a particular `--trace-config` as a shortcut. Concrete:

| verbosity | trace-config | Description |
|----------|-------------|----------------|
| `-v`        |                               | as it is now |
| `-vv`        |   `--trace-config ids,time`  |  good for parallel transfers, seeing delays |
| `-vvv`        |   `--trace-config ids,time,http/2,http/3,ftp`  |  all infos at the protocol level |
| `-vvvv`        |   `--trace-config all`  | give everything we got |

### History
In 5abfdc0140df097 (2008) the `-v` option was converted to a pure on/off switch. It used to do different things, but explicit command line options for that were introduced to replace it. Explicit options are good. Repeated `-v` should be about *usability* and not about *features*.

### Discussion

There are several soft issues here on how to *exactly* match increasing verbosity to features. Some points:

1. It is probably wise to leave a single `-v` as it is. *If* we think to change it, `ids` and `time` would probably be the only features worth to consider.
2. `-vvv` should contain all protocols we have tracing for. We might support tracing in more protocols in the future. We recently added it for ftp. So should we do for others.
3. `-vvvv` is awfully long. Maybe too long?

Is this a good idea? Which tracing shortcut would make your life easier?

### Update

For usability it is helpful to separate tracing at the *protocol* level from the *network* level. If one tries to find out if curl sends something wrong or the server's reply is unexpected, one looks at the protocol information (and `--trace-ascii %` is helpful here). In many cases, the lower layers like proxying, SSL and TCP work fine here and tracing output of those would just get in the way. That indicates that a level between just `-v` and `--trace-config all` is helpful.

**Comment by vszakats** [comment]
It seems to bring a new concept into the existing switches, with some room for confusion, e.g. what `-v -v`, or `-vfv` would do? Also `-vv` was already a valid option but would enable normal verbose logging. After this change it'd suddenly log more. What if there is one `-v` in `.curlrc` and another on the command-line? Also how to translate it to the `CURLOPT_VERBOSE` option, which so far was an on/off switch?

I like `-vv` in ssh, but in case of curl I'm siding with a different solution.

It's also _probably_ not something most users would need often.

Perhaps adding more short presets for `--trace-config`, following `all`? and/or replacing `--trace-config` with an eaiser to type name, like `--v<preset>`: `curl --vall` / `curl --v1` / `curl --v2` / `curl --v<preset>`?

**Reply by icing** [reply]
I  agree that this would change behaviour for people having `-v` in their `.curlrc` *and* adding it on the command line. Otoh, I think many people will try `-vvv` to get more tracing with curl without reading up on the man pages on how to achieve what they need. It is just a common usage pattern with many tools.

In that way, the change would meet users existing expectations. Documentation study is the last resort of the desperate... =)

**Comment by mschmitt** [comment]
Oh, so this is where the no-op `-vv` came from that I recently saw in a piece of older documentation. :-)

`-vv` as a mere shorthand for `--trace-config` that requires `--trace(-ascii)` and an output definition on top of it, might just not be what some users would expect from `-vv`. What I, for one, would expect, is `curl -vv` behaving like `curl --trace-ascii % --trace-config ...`.

I'm not a a heavy `--trace` user, though, and might be oblivious of the big picture.

**Reply by icing** [reply]
Maybe the output of `--trace-ascii %` should be part of the `-v` level where we trace protocol information (how many -v that may turn out to become)? That would make sense for me.

**Reply by mschmitt** [reply]
I'm not sure we're looking at this from the same angle. My perspective is that of me as mediocre systems guy who just wants to casually add more verbosity in the hope something useful might show up. (The inevitable parallel to `ssh -vvv` has already been drawn.) 

I only got in touch with `--trace` recently when trying to figure out a Kerberos-OIDC issue, not so much because I needed more verbose output but I rather needed more control where that output went, and specifying an output file was exactly the thing. I would expect `-vv...` to be more of an ad-hoc use case, hence my suggestion to implicitly redirect the trace to stderr.

**Reply by icing** [reply]
I agree with you. I think increasing `-v` should also automatically add the trace ascii output.

**Comment by bagder** [comment]
One detail to remember: because `-v` is treated as a boolean now, it supports `--no-verbose` to switch if off.

Even if we add an extra level of verbose per new `-v` and are therefore not handling it as a boolean anymore, we should keep support for `--no-verbose` to maintain old behavior for the rare instances where someone uses this.

**Comment by bagder** [comment]
I'm casually positive towards this proposal. A lot of users already expect `-v` to be working this way. Makes it slightly easier to get more debug info logged.

**Comment by icing** [comment]
Thanks for all the input. I made #13977 out of this. Any comments on the proposed change should then be in  the PR. Therefore I close this discussion.
